core/bloom.go

```diff
@@ -0,0 +1,358 @@
+	for i := 0; i < int(k); i++ {
+		opts.hashFns[i] = murmur3.SeedNew64(rand.Uint64())
+	}
+
+	// Calculate the number of bytes to be used
+	// 		bits = k * entries / ln(2)
+	//		bytes = bits * 8
+	bits := uint64(math.Ceil((k * float64(opts.capacity)) / ln2))
+	var bytes uint64
+	if bits%8 == 0 {
+		bytes = bits / 8
+	} else {
+		bytes = (bits / 8) + 1
+	}
+	opts.bits = bytes * 8
+
+	bitset := make([]byte, bytes)
+
+	return &Bloom{opts, bitset}
+}
+
+func (b *Bloom) info(name string) string {
+	info := ""
+	if name != "" {
+		info = "name: " + name + ", "
+	}
+	info += fmt.Sprintf("error rate: %f, ", b.opts.errorRate)
+	info += fmt.Sprintf("capacity: %d, ", b.opts.capacity)
+	info += fmt.Sprintf("total bits reserved: %d, ", b.opts.bits)
+	info += fmt.Sprintf("bits per element: %f, ", b.opts.bpe)
+	info += fmt.Sprintf("hash functions: %d", len(b.opts.hashFns))
+
+	return info
+}
+
+// add adds a new entry for `value` in the filter. It hashes the given
+// value and sets the bit of the underlying bitset. Returns "-1" in
+// case of errors, "0" if all the bits were already set and "1" if
+// atleast 1 new bit was set.
+func (b *Bloom) add(value string) ([]byte, error) {
+	// We're sure that empty values will be handled upper functions itself.
+	// This is just a property check for the bloom struct.
+	if value == "" {
+		return RESP_MINUS_1, errEmptyValue
+	}
+
+	// Get the indexes where bits are supposed to be set
+	indexes, err := b.opts.getIndexes(value)
+	if err != nil {
+		fmt.Println("error in getting indexes for value:", value, "err:", err)
+		return RESP_MINUS_1, errUnableToHash
+	}
+
+	// Set the bits and keep a count of already set ones
+	count := 0
+	for _, v := range indexes {
+		if isBitSet(b.bitset, int(v)) {
+			count++
+		} else {
+			setBit(b.bitset, int(v))
+		}
+	}
+
+	if count == len(indexes) {
+		// All the bits were already set, return 0 in that case.
+		return RESP_ZERO, nil
+	}
+
+	return RESP_ONE, nil
+}
+
+// exists checks if the given `value` exists in the filter or not.
+// It hashes the given value and checks if the bits are set or not in
+// the underlying bitset. Returns "-1" in case of errors, "0" if the
+// element surely does not exist in the filter, and "1" if the element
+// may or may not exist in the filter.
+func (b *Bloom) exists(value string) ([]byte, error) {
+	// We're sure that empty values will be handled upper functions itself.
+	// This is just a property check for the bloom struct.
+	if value == "" {
+		return RESP_MINUS_1, errEmptyValue
+	}
+
+	// Get the indexes where bits are supposed to be set
+	indexes, err := b.opts.getIndexes(value)
+	if err != nil {
+		fmt.Println("error in getting indexes for value:", value, "err:", err)
+		return RESP_MINUS_1, errUnableToHash
+	}
+
+	// Check if all the bits at given indexes are set or not
+	// Ideally if the element is present, we should find all set bits.
+	for _, v := range indexes {
+		if !isBitSet(b.bitset, int(v)) {
```

Comment:
`int` -> see if we can converge to one integer variant. We can add that limitation of `32` or `64`.